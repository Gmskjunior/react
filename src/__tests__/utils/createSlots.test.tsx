import React from 'react'
import 'babel-polyfill'
import {render, waitFor} from '@testing-library/react'
import createSlots from '../../utils/create-slots'

// setup a component with slots
const {useSlots, Slot} = createSlots<'One' | 'Two' | 'Three'>()

const ComponentWithSlots: React.FC<React.PropsWithChildren<unknown>> = ({children}) => {
  const {slots, SlotsProvider} = useSlots()

  return (
    <SlotsProvider>
      <div>
        {slots.One}
        <span>
          {children} {slots.Two} {slots.Three}
        </span>
      </div>
    </SlotsProvider>
  )
}
const SlotItem1: React.FC<React.PropsWithChildren<unknown>> = ({children}) => <Slot name="One">{children}</Slot>
const SlotItem2: React.FC<React.PropsWithChildren<unknown>> = ({children}) => <Slot name="Two">{children}</Slot>

describe('ComponentWithSlots', () => {
  it('renders all slots', async () => {
    const component = render(
      <ComponentWithSlots>
        <SlotItem1>first</SlotItem1>
        <SlotItem2>second</SlotItem2>
        free form
      </ComponentWithSlots>
    )

    await waitFor(() => component.getByText('first'))
    expect(component.container).toMatchSnapshot()
  })

  it('renders without any slots', async () => {
    const component = render(<ComponentWithSlots>free form</ComponentWithSlots>)
    expect(component.container).toMatchSnapshot()
  })

  it('renders with just one slot', async () => {
    const component = render(
      <ComponentWithSlots>
        <SlotItem1>first</SlotItem1>
        free form
      </ComponentWithSlots>
    )
    expect(component.container).toMatchSnapshot()
  })

  it('works with custom context', async () => {
    const Context = React.createContext({salutation: 'not found'})

    const ContextReadingSlot = () => {
      const {salutation} = React.useContext(Context)
      return <SlotItem1>{salutation}</SlotItem1>
    }

    const component = render(
      <Context.Provider value={{salutation: 'hello'}}>
        <ComponentWithSlots>
          <ContextReadingSlot />
          free form
        </ComponentWithSlots>
      </Context.Provider>
    )
    expect(component.container).toMatchSnapshot()
  })
})
