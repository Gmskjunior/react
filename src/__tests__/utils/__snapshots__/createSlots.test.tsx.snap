// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentWithSlots renders all slots 1`] = `
<div>
  <div>
    first
    <span>
      free form
       
      second
       
    </span>
  </div>
</div>
`;

exports[`ComponentWithSlots renders with just one slot 1`] = `
<div>
  <div>
    first
    <span>
      free form
       
       
    </span>
  </div>
</div>
`;

exports[`ComponentWithSlots renders without any slots 1`] = `
<div>
  <div>
    <span>
      free form
       
       
    </span>
  </div>
</div>
`;

exports[`ComponentWithSlots works with custom context 1`] = `
<div>
  <div>
    hello
    <span>
      free form
       
       
    </span>
  </div>
</div>
`;
